docs/upgrade.md

```diff
@@ -89,6 +89,231 @@ process, for example:
     dpkg -i matrix-synapse-py3_1.3.0+stretch1_amd64.deb
     ```
 
+# Upgrading to v1.61.0
+
+## New signature for the spam checker callback `user_may_join_room`
+
+The previous signature has been deprecated.
+
+Whereas `user_may_join_room` callbacks used to return `bool`, they should now return `Union["synapse.module_api.NOT_SPAM", "synapse.module_api.errors.Codes"]`.
+
+This is part of an ongoing refactoring of the SpamChecker API to make it less ambiguous and more powerful.
```

Comment:
I'm not sure I feel super happy including this line in the user-facing upgrade notes, because it feels to me like the refactoring is internal context that the user doesn't need to know and the "less ambiguous and more powerful" part is a bit subjective.

Reply:
Well, I'd argue that sometimes returning `True` to mean "it's spam" and sometimes `True` to mean "it's not spam" is definitely more ambiguous :)

But we can live without that sentence.